### PR TITLE
error logging with fmt.Errorf

### PR DESCRIPTION
**Pull request by j4k4**
Since golang 1.13 there is fmt.Errorf to create and wrap errors which allows the usual msg and args pattern to format and create errors.  
We use this functionality now to create the final error which is used now on logging any error.

### CODE REVIEW

File: pkg/cfg/config.go [hunk 1]
```diff
@@ -177,8 +177,8 @@ func (c *config) GetMsiSlice(key string, optionalDefault ...[]map[string]interfa
 	var reflectValue = reflect.ValueOf(data)
 
 	if reflectValue.Kind() != reflect.Slice {
-		err = fmt.Errorf("can not cast value %v[%T] of key %s to []map[string]interface{}", data, data, key)
-		c.err(err, err.Error())
+		err = fmt.Errorf("can not cast value %v[%T] of key %s to []map[string]interface{}: %w", data, data, key, err)
+		c.err(err.Error())
```
Comment (ajscholl) [inline pkg/cfg/config.go:181]: Shouldn't this be `c.err("can not cast value %v[%T] of key %s to []map[string]interface{}: %w", data, data, key, err)`?

File: pkg/cfg/config.go [hunk 2]
```diff
@@ -620,7 +619,8 @@ func (c *config) unmarshalStruct(key string, output interface{}, additionalDefau
 		settings, err := c.settings.Get(key).Map()
 
 		if err != nil {
-			c.err(fmt.Errorf("value is not of type map[string]interface{}: %w", err), "can not get settings for key: %s", key)
+			err = fmt.Errorf("value is not of type map[string]interface{}: %w", err)
```
Comment (ajscholl) [inline pkg/cfg/config.go:622]: Is this line correct? What is `value` in this context?

File: pkg/cfg/error.go [hunk 3]
```diff
@@ -1,37 +1,37 @@
...
-	err = errors.Wrapf(err, msg, args...)
+func PanicErrorHandler(msg string, args ...interface{}) {
+	err := fmt.Errorf(msg, args...)
 	panic(err)
 }
 
 func LoggerErrorHandler(logger Logger) ErrorHandler {
-	return func(err error, msg string, args ...interface{}) {
-		logger.Error(err, msg, args...)
+	return func(msg string, args ...interface{}) {
```
Comment (ajscholl) [inline pkg/cfg/error.go:27]: So this whole could be replaced by `return logger.Error`?

File: pkg/cfg/error.go [hunk 4]
```diff
@@ -1,37 +1,37 @@
...
+		logger.Error(msg, args...)
 	}
 }
 
 func SentryErrorHandler(sentry Sentry) ErrorHandler {
-	return func(err error, msg string, args ...interface{}) {
-		err = errors.Wrapf(err, msg, args...)
+	return func(msg string, args ...interface{}) {
+		err := fmt.Errorf(msg, args...)
```
Comment (ajscholl) [inline pkg/cfg/error.go:34]: I wonder... `errors.Wrapf` adds a stack trace somehow... do we need this? Does sentry add its own stack trace?

File: pkg/db-repo/change_history_manager.go [hunk 5]
```diff
@@ -224,7 +224,7 @@ func (c *changeHistoryManager) execute() error {
 		if err != nil {
 			c.logger.WithFields(mon.Fields{
 				"sql": statement,
-			}).Error(err, "")
+			}).Error(": %w", err)
```
Comment (ajscholl) [inline pkg/db-repo/change_history_manager.go:227]: This looks like it could need a proper error message

File: pkg/db-repo/notification_repo.go [hunk 6]
```diff
@@ -77,10 +77,7 @@ func (r *notifyingRepository) doCallback(ctx context.Context, callbackType strin
 	}
 
 	if len(errors) > 0 {
-		err := fmt.Errorf("there were %v errors during execution of the callbacks for create", len(errors))
-		logger.Error(err, err.Error())
-
-		return err
+		return fmt.Errorf("there were %v errors during execution of the callbacks for create", len(errors))
```
Comment (ajscholl) [inline pkg/db-repo/notification_repo.go:80]: Maybe we should replace the whole error by a multierror? I for my part hate it if I know there were 7 errors, but don't know what the errors were...

File: pkg/dispatcher/repository_db.go [hunk 7]
```diff
@@ -36,7 +36,7 @@ func (r Repository) Create(ctx context.Context, value db_repo.ModelBased) error
 
 	for _, err := range errs {
 		if err != nil {
-			r.logger.Error(err, "error on "+db_repo.Create+" for event "+eventName)
+			r.logger.Error("error on "+db_repo.Create+" for event : %w"+eventName, err)
```
Comment (ajscholl) [inline pkg/dispatcher/repository_db.go:39]: This file seems to mix format strings and string concatenation

File: pkg/mon/mocks/factory.go [hunk 8]
```diff
@@ -38,38 +38,32 @@ func NewLoggerMockedUntilLevel(level string) *Logger {
...
+	mockLoggerMethod(logger, "Info", mon.Info, levelIndex >= levelMap[mon.Info])
+	mockLoggerMethod(logger, "Warn", mon.Warn, levelIndex >= levelMap[mon.Warn])
+	mockLoggerMethod(logger, "Error", mon.Error, levelIndex >= levelMap[mon.Error])
 
 	return logger
 }
 
 func inspectLogFunction(level string, allowed bool) func(args mock.Arguments) {
 	return func(args mock.Arguments) {
 		// ensure we use formatting markers exactly when using a method with formatting
```
Comment (ajscholl) [inline pkg/mon/mocks/factory.go:51]: I think we can delete this comment, it is no longer valid

File: pkg/cfg/config.go (hunk 1 above)
Comment (j4k4) [reply pkg/cfg/config.go:181]: indeed

File: pkg/cfg/config.go (hunk 2 above)
Comment (j4k4) [reply pkg/cfg/config.go:622]: that wrapping can be dropped, as the `Map()` call already returns an appropriate error

File: pkg/cfg/error.go (hunk 4 above)
Comment (j4k4) [reply pkg/cfg/error.go:34]: Luckily the `SentryErrorHandler` is not used anywhere, so we don't have to care about this here..  
But in general the sentry stuff needs some overhaul, as it uses quite some old error packages.  
This is planned for the incoming new sentry handler for the incoming new logger package..

File: pkg/db-repo/change_history_manager.go (hunk 5 above)
Comment (j4k4) [reply pkg/db-repo/change_history_manager.go:227]: added the message from the error return..

File: pkg/db-repo/notification_repo.go (hunk 6 above)
Comment (j4k4) [reply pkg/db-repo/notification_repo.go:80]: using multierror now

File: pkg/dispatcher/repository_db.go (hunk 7 above)
Comment (j4k4) [reply pkg/dispatcher/repository_db.go:39]: fixed

File: pkg/mon/mocks/factory.go (hunk 8 above)
Comment (j4k4) [reply pkg/mon/mocks/factory.go:51]: done

File: pkg/db-repo/notification_repo.go [hunk 9]
```diff
@@ -65,22 +66,19 @@ func (r *notifyingRepository) doCallback(ctx context.Context, callbackType strin
...
 		}
 	}
 
-	if len(errors) > 0 {
-		err := fmt.Errorf("there were %v errors during execution of the callbacks for create", len(errors))
-		logger.Error(err, err.Error())
-
-		return err
+	if errors != nil {
+		return fmt.Errorf("there were errors during execution of the callbacks for create: %w", errors)
```
Comment (ajscholl) [inline pkg/db-repo/notification_repo.go:81]: Actually they are not the callbacks for create, but for `callbackType`, right?

**Review by ajscholl — APPROVED** [review]

**Pull request merged**